Add Product metadata and addon support to WCProductModel

ThomazFB opened the pull request:
Summary
==========
Fixes issue https://github.com/woocommerce/woocommerce-android/issues/4405 by extending the `WCProductModel` to support the Metadata group of data and filtering the Add-ons section with the desired model mapping.

How to Test
==========
**Prerequisites:** Have your site with the [add ons plugin](https://woocommerce.com/products/product-add-ons/) installed, configured, and with at least one product with add-ons. 

- Launch the example app.
- Navigate to the Product section inside the Woo area and select the `Fetch Product addons` button.
- See that add-ons info is being printed in the console.

Screenshots
==========
| Step 1  | Step 2 | Step 3 | Step 4 |
| ------------- | ------------- | ------------- | ------------- |
| ![Screenshot_20210722_184253](https://user-images.githubusercontent.com/5920403/126714334-03f0a65c-1b82-446c-9b6c-e4b6cb4ac169.png) | ![Screenshot_20210722_184258](https://user-images.githubusercontent.com/5920403/126714339-80f6b7cf-4492-4f20-83b3-77d91d79ee10.png) | ![Screenshot_20210722_184306](https://user-images.githubusercontent.com/5920403/126714344-21886a57-6b06-4305-ab15-539c6b3867da.png) | ![Screenshot_20210722_184321](https://user-images.githubusercontent.com/5920403/126714349-76992eee-2746-4d67-a359-4f6ea538b339.png) |

ThomazFB committed:
Create initial class scaffold for Product Addons

ThomazFB committed:
Create Product Addon base API response

ThomazFB committed:
Create Product Addon Type API response

ThomazFB committed:
Refactor ProductAddonApiResponse to use enums to define internal types

ThomazFB committed:
Set remaining ApiResponses for ProductAddon

ThomazFB committed:
Create database metadata field for WCProductModel

ThomazFB committed:
Set metadata field on ProductApiResponse

ThomazFB committed:
Add metadata field support to WCProductModel

ThomazFB committed:
Add addons metadata JSON mapping inside WCProductModel

ThomazFB committed:
Add serialization types to ProductAddonApiResponse

ThomazFB committed:
Add serialization types to ProductAddonOptionApiResponse

ThomazFB committed:
Set metadata to addon serialization and parsing inside WCProductModel

ThomazFB committed:
Remove unnecessary scaffold

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge branch 'develop' into woo/4405-product-addon-metadata

ThomazFB committed:
Rename and relocate Addons models

ThomazFB committed:
Fix ProductAddon parsing

ThomazFB committed:
Extract Json file serialization method to JsonExt

ThomazFB committed:
Add Product metadata serialization test

ThomazFB committed:
Remove empty space

ThomazFB committed:
Add more tests to WCProductModelTest

ThomazFB committed:
Improve addons tests

ThomazFB committed:
Fix redundant JsonLoader object structure

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix lint issues (again)

ThomazFB committed:
Add Product addon visualization to example app

ThomazFB committed:
Add error handling message when no addons found

ThomazFB committed:
Fix lint issues (Again the second)

ThomazFB commented:
Hey @wzieba, I'm requesting your review from this PR since you will be joining the Add-ons product soon. Since this PR introduces the first data parsing for our future work, reviewing it may be a good entry point for technical context.

ThomazFB committed:
Extract addon log code to method

wzieba commented on example/src/main/java/org/wordpress/android/fluxc/example/ui/products/WooProductsFragment.kt:
```diff
@@ -371,6 +377,26 @@ class WooProductsFragment : Fragment() {
+                        coroutineScope.launch {
+                            wcProductStore.fetchProductListSynced(site, listOf(id))
+                                    .takeUnless { it.isNullOrEmpty() }
+                                    ?.first()?.addons
+                                    ?.logAddons()
+                        }
+                    } ?: prependToLog("No valid remoteOrderId defined...doing nothing")
```
❓ If we have a log for `not valid remoteOrderId`, maybe we could have one for not finding `addons`?

ThomazFB replied:
Actually, we do have a log for this scenario, but it's inside the `logAddons()` method, do you think it's too hidden and should be handled at the method caller instead of inside the method itself?

Here's how implemented:
```
private fun Array<WCProductAddonModel>?.logAddons() {
         this?.forEachIndexed { index, addon ->
             prependToLog(addon.description?.let { "description: $it" }.orEmpty())
             prependToLog(addon.name?.let { "name: $it" }.orEmpty())
             prependToLog("========== Product Add-on #$index =========")
         } ?: prependToLog("No addons found for this product ID")
     }
```

wzieba replied:
Oh missed that, thanks! But I think it might not work, please correct me if I'm wrong: 

if [`addons` are not found in payload, we return an emptyList](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2066/files#diff-62817f824846374ea4f2a7620ebef16d42efdb34f1990600f9c10b215c9c4f39R113). Next, in the snippet you've provided we log `No addons found for this product ID` for null `Array<WCProductAddonModel>` but it never happens.

I think `null`, if there are no add-ons in payload, is a valid state. I'd suggest to change `WCProductModel#addons` to return `null` if there are no add-ons, WDYT?

ThomazFB replied:
Great catch @wzieba! You're 100% correct, looks like I got fooled by my own safety measure 😅. I agree with your suggestion, it's way better to use a nullable state to represent data failure instead of an empty array. I've submitted the change right here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2066/commits/65f612f39055b57ecae23a5ea8fd62141a9cff3e.

Let me know what you think and thanks for your time reviewing the PR!

wzieba replied:
Thanks for change! I've added one more comment here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2066/files#r676488729

wzieba reviewed (approved):
Thanks for adding tests! LGTM, one minor comment about logging - I think it could be a useful one to have log for getting no add-ons when requested.

About testing: at first I didn't get add-on as I've added a global one and global add-ons will be handled differently that added here support for single product add-on.

ThomazFB commented:
Thanks for your review @wzieba!

> I think it could be a useful one to have log for getting no add-ons when requested.

Nice! I replied your comment, thanks for point this out!

> About testing: at first I didn't get add-on as I've added a global one and global add-ons will be handled differently that added here support for single product add-on.

I see! Awesome feedback! I'll try to differentiate better for the next PRs if the test is about Product, Order or Global add-ons, thank you for bringing this up.

ThomazFB committed:
Adapt addons field to be nullable in case of no data acquired

ThomazFB committed:
Adjust tests to follow addons field nullability rule

hichamboushaba commented:
I have a remark about this @ThomazFB 🙂, we introduced a reusable class to represent all metadatas of WC objects https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e1fc3eb87caa0a824d512dc5c902a6ed9e0d09a2/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/WCMetaData.kt, WDYT about reusing it instead of adding the new class `ProductMetadata`?

wzieba commented on example/src/main/java/org/wordpress/android/fluxc/example/ui/products/WooProductsFragment.kt:
```diff
@@ -371,6 +377,26 @@ class WooProductsFragment : Fragment() {
+                        coroutineScope.launch {
+                            wcProductStore.fetchProductListSynced(site, listOf(id))
+                                    .takeUnless { it.isNullOrEmpty() }
+                                    ?.first()?.addons
+                                    ?.logAddons()
```
I think here we also have to adjust the code: 
```suggestion
                                    ?.first()?.addons
                                    .logAddons()
```

Otherwise, when `addons` is `null`, `logAddons()` method won't be executed. I've tested it after removing `?` and now it seems to work 🙂:

![image](https://user-images.githubusercontent.com/5845095/126976021-2443fb4c-b9e7-4b3c-9664-b3df52a657ae.png)

ThomazFB replied:
This was a lack of attention of mine, sorry, it's fixed here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2066/commits/096b4455d600ace1465357ab5e896cb7505a9afd

ThomazFB committed:
Remove ProductMetadata model and take advantage of WCMetaData class

ThomazFB committed:
Remove unnecessary nullable check for null safe function

ThomazFB committed:
Fix serialization for WCMetaData.value

ThomazFB commented:
@wzieba, once again, thanks for catching issues, amazing review work! The PR is ready for another round.

> WDYT about reusing it instead of adding the new class ProductMetadata?

Hey @hichamboushaba, thanks for letting me know about this, I wasn't aware that metadata is present in the same data format for other objects, great tip! I've deleted the internal `ProductMetadata` and adjusted the `WCProductModel` to take advantage of the existing `WCMetaData` 👍.

ThomazFB committed:
Add the Unknown Type for every Addon enum

ThomazFB committed:
Remove the Unknown Type of every Addon enum

ThomazFB committed:
Merge branch 'develop' into woo/4405-product-addon-metadata

wzieba commented:
Thanks @ThomazFB ! It all looks great now! Also thank you @hichamboushaba for your suggestion 🙌 .
